### PR TITLE
Multi adapter RL (MARL) - a single model for RM & Value Head

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR introduces a new concept (initiated by @pacman100 & @philschmid) - were you would use a single base model to apply both RM & PPO leveraging multiple adapters.

image

Users needs first to train RM adapters using PEFT (script will be provided soon), then push these adapters either on the Hub or save them locally.

After that, just load the RM adapter using the flag `reward_modeling_adapter_id` when creating a `PPOTrainer`. Note that this feature is exclusive the fine-tuning PPO with adapters 

```python
model = AutoModelForCausalLMWithValueHead.from_pretrained(
    model_name,
    load_in_8bit=True,
    device_map={"": 0},
    peft_config=lora_config,
    reward_modeling_adapter_id=rm_adapter_id,
)
```

Then at rollout time call `ppo_trainer.model.compute_reward_score` to compute the raw reward scores of the RM. That's all you need!

With that you could fine-tune llama-7b on a single Nvidia T4 using RLHF (with 8bit base model loading) - meaning this would unlock many possibility for users and practitioneers.

## TODOs

- [x] nice CI tests
- [x] precise docs

cc @lvwerra @lewtun

### CODE REVIEW

File: docs/source/0_abstraction_rl.mdx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+...
+trainer = PPOTrainer(
+    model=model,
+    ...
+)
+
+...
+```
+Then inside your PPO training loop, call `compute_reward_score` method by accessing to the `model` attribute from `ppo_trainer`.
```
Comment (lewtun) [inline docs/source/0_abstraction_rl.mdx:51]:
```suggestion
Then inside your PPO training loop, call the `compute_reward_score` method by accessing to the `model` attribute from `PPOTrainer`.
```

File: docs/source/0_abstraction_rl.mdx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+If you are familiar with the `peft` library, you know that you can use multiple adapters inside the same model. What you can do is to train multiple adapters on the same base model to fine-tune on different policies. 
+In this case, you want to have a control on the adapter name you want to activate back, after retrieving the reward. For that, simply pass the appropriate `adapter_name` to `ppo_adapter_name` argument when calling `compute_reward_score`.
+
+```python
+adapter_name_policy_1 = "policy_1"
+rewards = trainer.model.compute_reward_score(**inputs, ppo_adapter_name=adapter_name_policy_1)
+...
+```
+
+### Using 8-bit base models
```
Comment (lewtun) [inline docs/source/0_abstraction_rl.mdx:70]: Should we also add a note on INT4 now that it's out? E.g. call this section "Using 4-bit and 8-bit base models"

File: examples/0-abstraction-hh-rlhf/rl_finetuning.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+model = AutoModelForCausalLMWithValueHead.from_pretrained(
+    model_name,
+    load_in_8bit=True,
+    device_map={"": 0},
+    peft_config=lora_config,
+    reward_modeling_adapter_id=rm_adapter_id,
+)
+tokenizer = LlamaTokenizer.from_pretrained(model_name)
+
+tokenizer.add_special_tokens(
```
Comment (lewtun) [inline examples/0-abstraction-hh-rlhf/rl_finetuning.py:69]: Same comment here about not needing this anymore (it might depend on the pretrained checkpoint you're using though!)

File: trl/models/modeling_base.py [hunk 4]
```diff
@@ -258,6 +267,14 @@ class and the arguments that are specific to trl models. The kwargs
         if is_resuming_training:
             model.post_init(state_dict=state_dict)
 
+        if not is_peft_model and reward_modeling_adapter_id is not None:
```
Comment (lewtun) [inline trl/models/modeling_base.py:270]: Do we also need a check that the base model is the same architecture that the reward model adapter (ie check that `llama-7b` is matched to `ybelkada/llama-7b-dummy-rm`

File: trl/trainer/utils.py [hunk 5]
```diff
@@ -226,3 +227,13 @@ def __iter__(self):
                     "input_ids": torch.LongTensor(example),
                     "labels": torch.LongTensor(example),
                 }
+
+
+class PeftSavingCallback(TrainerCallback):
```
Comment (lewtun) [inline trl/trainer/utils.py:232]: Should we have an integration test to check this does what we expect (or perhaps you already have it included in the tests?)

File: examples/0-abstraction-hh-rlhf/reward_modeling.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    # TODO: make it more userfriendly
+    device_map = {"": 0}
+
+    model = AutoModelForSequenceClassification.from_pretrained(
+        args.model_name, quantization_config=bnb_config, device_map=device_map
+    )
+    model = prepare_model_for_int8_training(model)
+
+    peft_config = LoraConfig(
+        lora_alpha=32, lora_dropout=0.05, bias="none", task_type="SEQ_CLS", modules_to_save=["score"]
```
Comment (lewtun) [inline examples/0-abstraction-hh-rlhf/reward_modeling.py:129]: Maybe we can explain why we need to also save the `score` module?

**Review by lewtun — APPROVED** [review]
Thanks a lot for this really excellent feature @younesbelkada - super cool to see memory getting reduced dramatically with this approach!

I left a few minor comments / suggestions, but the core implementation looks sound to me

File: examples/0-abstraction-hh-rlhf/rl_finetuning.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+from datasets import load_dataset
+from peft import LoraConfig
+from tqdm import tqdm
+from transformers import LlamaTokenizer
+
+from trl import AutoModelForCausalLMWithValueHead, PPOConfig, PPOTrainer
+from trl.core import LengthSampler
+
+
+rm_adapter_id = "ybelkada/llama-7b-dummy-rm"
```
Comment (lewtun) [inline examples/0-abstraction-hh-rlhf/rl_finetuning.py:24]: Have you run an experiment just using sentiment analysis (i.e. train the RM on IMDB) as a sanity check everything works?

**Comment by mnoukhov** [comment]
This is a great idea. I'd be interested in extending it to also make the value function use separate LoRA weights.

Would it also make sense to use python's `contextlib` to enable different LoRA weights within a scope? Something like 
```
with pretrained_model.reward_adapter():
     # calls pretrained model with the reward adapter weights
     pretrained_model.forward(...)
```

File: docs/source/0_abstraction_rl.mdx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+## Requirements
+
+You just need to install `peft` and optionally install `bitsandbytes` as well if you want to go for 8bit base models, for more memory efficient finetuning.
+
+## Summary
+
+You need to address this approach in three stages that we summarize as follows:
+
+1- Train a base model on the target domain (e.g. `imdb` dataset) - this is the Supervised Fine Tuning stage - it can leverage the `SFTTrainer` from TRL.
+2- Train a reward adapter model using `peft`. We show an example of leveraging the `RewardTrainer` from TRL in [this example](TODO)
```
Comment (younesbelkada) [inline docs/source/0_abstraction_rl.mdx:14]:
```suggestion
2- Train a reward adapter model using `peft`. We show an example of leveraging the `RewardTrainer` from TRL in [this example](https://github.com/lvwerra/trl/tree/main/examples/0-abstraction-RL/reward_modeling.py)
```

File: docs/source/0_abstraction_rl.mdx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+1- Train a base model on the target domain (e.g. `imdb` dataset) - this is the Supervised Fine Tuning stage - it can leverage the `SFTTrainer` from TRL.
+2- Train a reward model using `peft`. This is required in order to re-use the adapter during the RL optimisation process (step 3 below). We show an example of leveraging the `RewardTrainer` from TRL in [this example](https://github.com/lvwerra/trl/tree/main/examples/0-abstraction-RL/reward_modeling.py)
+3- Fine tune new adapters on the base model using PPO and the reward adapter. ("0 abstraction RL")
+
+Make sure to use the same model (i.e. same architecure and same weights) for the stages 2 & 3. 
+
+## Quickstart
+
+Let us assume you have trained your reward adapter on `llama-7b` model using `RewardTrainer` and pushed the weights on the hub under `ybelkada/llama-7b-hh-rm-adapter`.
```
Comment (younesbelkada) [inline docs/source/0_abstraction_rl.mdx:21]:
```suggestion
Let us assume you have trained your reward adapter on `llama-7b` model using `RewardTrainer` and pushed the weights on the hub under `trl-lib/llama-7b-hh-rm-adapter`. 
```

File: docs/source/0_abstraction_rl.mdx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+Make sure to use the same model (i.e. same architecure and same weights) for the stages 2 & 3. 
+
+## Quickstart
+
+Let us assume you have trained your reward adapter on `llama-7b` model using `RewardTrainer` and pushed the weights on the hub under `ybelkada/llama-7b-hh-rm-adapter`. 
+When doing PPO, before passing the model to `PPOTrainer` create your model as follows:
+
+```python
+model_name = "llama-7b"
+rm_adapter_id = "ybelkada/llama-7b-hh-rm-adapter"
```
Comment (younesbelkada) [inline docs/source/0_abstraction_rl.mdx:26]:
```suggestion
rm_adapter_id = "trl-lib/llama-7b-hh-rm-adapter"
```

File: docs/source/0_abstraction_rl.mdx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+Make sure to use the same model (i.e. same architecure and same weights) for the stages 2 & 3. 
+
+## Quickstart
+
+Let us assume you have trained your reward adapter on `llama-7b` model using `RewardTrainer` and pushed the weights on the hub under `ybelkada/llama-7b-hh-rm-adapter`. 
+When doing PPO, before passing the model to `PPOTrainer` create your model as follows:
+
+```python
+model_name = "llama-7b"
```
Comment (younesbelkada) [inline docs/source/0_abstraction_rl.mdx:25]:
```suggestion
model_name = "huggyllama/llama-7b"
```

File: docs/source/0_abstraction_rl.mdx [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+...
+```
+
+### Using 8-bit base models
+
+For more memory efficient fine-tuning, you can load your base model in 8-bit while keeping the adapters in the default precision (float32).
+Just pass the appropriate arguments (i.e. `load_in_8bit=True`) to `AutoModelForCausalLMWithValueHead.from_pretrained` as follows (assuming you have installed `bitsandbytes`):
+```python
+model_name = "llama-7b"
+rm_adapter_id = "ybelkada/llama-7b-hh-rm-adapter"
```
Comment (younesbelkada) [inline docs/source/0_abstraction_rl.mdx:76]:
```suggestion
rm_adapter_id = "trl-lib/llama-7b-hh-rm-adapter"
```

File: trl/models/modeling_base.py [hunk 13]
```diff
@@ -107,6 +110,7 @@ class and the arguments that are specific to trl models. The kwargs
         """
         if kwargs is not None:
             peft_config = kwargs.pop("peft_config", None)
+            reward_modeling_adapter_id = kwargs.pop("reward_modeling_adapter_id", None)
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/models/modeling_base.py:113]: Agreed! Let's use `reward_adapter`!

File: trl/trainer/utils.py (hunk 5 above)
Comment (younesbelkada) [reply trl/trainer/utils.py:232]: yes this should be already tested in the sft trainer tests, I just moved it in utils so that the RM Trainer can also benefit from it!

File: trl/models/modeling_base.py [hunk 14]
```diff
@@ -365,6 +382,81 @@ def post_init(self, *args, **kwargs):
...
+                    "Could not find adapter model in the Hub, make sure you have the correct adapter model id."
+                )
+        else:
+            local_filename = filename
+
+        adapter_state_dict = torch.load(local_filename, map_location="cpu")
+        rm_adapter_peft_config = LoraConfig.from_pretrained(adapter_model_id)
+
+        for score_name_candidate in self.supported_rm_modules:
+            if any([score_name_candidate in name for name in adapter_state_dict.keys()]):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/models/modeling_base.py:407]: I think this is because we always just find a single head among the list of possible score head naming. Once we found it, we can safely break. Added a comment!

File: examples/0-abstraction-hh-rlhf/reward_modeling.py (hunk 6 above)
Comment (younesbelkada) [reply examples/0-abstraction-hh-rlhf/reward_modeling.py:126]: added some comments above!

**Comment by mnoukhov** [comment]
The `torch.no_grad` isn't necessary but currently the lora parameters of the reward model still require grad. I think if you add `inference_mode = True` to the `rm_adapter_peft_config` then it will set those parameters to `requires_grad = False` and can do without `torch.no_grad`

**Review by lvwerra — APPROVED** [review]
